src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs

```diff
@@ -701,6 +767,48 @@ public virtual int ReadByte()
             return r == 0 ? -1 : oneByteArray[0];
         }
 
+        public void ReadExactly(Span<byte> buffer) =>
+            _ = ReadAtLeastCore(buffer, buffer.Length, throwOnEndOfStream: true);
+
+        public void ReadExactly(byte[] buffer, int offset, int count)
+        {
+            ValidateBufferArguments(buffer, offset, count);
+
+            _ = ReadAtLeastCore(buffer.AsSpan(offset, count), count, throwOnEndOfStream: true);
+        }
+
+        public int ReadAtLeast(Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream = true)
+        {
+            ValidateReadAtLeastArguments(buffer.Length, minimumBytes);
+
+            return ReadAtLeastCore(buffer, minimumBytes, throwOnEndOfStream);
+        }
+
+        // No argument checking is done here. It is up to the caller.
+        private int ReadAtLeastCore(Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream)
+        {
+            int totalRead = 0;
+            do
+            {
+                int read = Read(buffer.Slice(totalRead));
+                if (read == 0)
+                {
+                    if (throwOnEndOfStream)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
```

Comment:
I wonder why the method is called `ThrowEndOfFileException`, yet it throws an `EndOfStreamException` internally.

BTW, this is unrelated to your changes, but TIL of the existence of `ThrowHelper`. But it is in S.P.CoreLib, I would've liked to use it inside the new tar assembly. Would it make sense to request moving it to libraries/Common for wider usage? Or is it in S.P.CoreLib for optimization purposes I am not aware of?

Reply:
> I wonder why the method is called ThrowEndOfFileException, yet it throws an EndOfStreamException internally.

Agreed. It is a bit weird, but I wasn't going to mess with it here.

> But it is in S.P.CoreLib, I would've liked to use it inside the new tar assembly. Would it make sense to request moving it to libraries/Common for wider usage? Or is it in S.P.CoreLib for optimization purposes I am not aware of?

We have a few of them in across different assemblies. Since each assembly typically has its own exceptions to be thrown, I'm not sure it makes sense to reuse them across assemblies. We have one common (partial) one for ArgNull:

https://github.com/dotnet/runtime/blob/20cd393c1e577f0ff53cdc3a00edc3bcfc68ab7f/src/libraries/Common/src/System/ThrowHelper.cs#L12

See https://source.dot.net/#q=ThrowHelper for all of them.